src/goto-symex/symex_complexity.h

```diff
@@ -0,0 +1,155 @@
+/// the complexity of the loop as a whole is acceptable.
+class symex_complexityt
+{
+public:
+  symex_complexityt() = default;
+
+  explicit symex_complexityt(message_handlert &logger) : log(logger)
+  {
+  }
+
+  symex_complexityt(message_handlert &logger, const optionst &options)
+    : log(logger)
+  {
+    std::size_t limit = options.get_signed_int_option("symex-complexity-limit");
+    if((complexity_active = limit > 0))
+    {
+      // This gives a curve that allows low limits to be rightly restrictive,
+      // while larger numbers are very large to represent what you're asking.
+      max_complexity = static_cast<std::size_t>((limit * limit) * 75);
+
+      std::size_t failed_children_limit =
+        options.get_signed_int_option("symex-complexity-failed-children-limit");
+      std::size_t unwind = options.get_signed_int_option("unwind");
+
+      // In order of priority:
+      //  * explicit limit
+      //  * inferred limit from unwind
+      //  * best limit we can apply with very little information.
+      if(failed_children_limit > 0)
+        max_loops_complexity = failed_children_limit;
+      else if(unwind > 0)
+        max_loops_complexity = std::max(
+          static_cast<std::size_t>(std::floor(unwind / 3)),
+          static_cast<std::size_t>(1));
+      else
+        max_loops_complexity = limit;
+    }
+  }
+
+protected:
+  mutable messaget log;
+
+  /// Is the complexity module active, usually coincides with a max_complexity
+  /// value above 0.
+  bool complexity_active = false;
+
+  /// Functions called when the heuristic has been violated. Can perform
+  /// any form of branch, state or full-program transformation.
+  std::vector<symex_heuristic_transformationt> violation_transformations;
+
+  /// Default heuristic transformation. Adds unconditional false guards to
+  /// state.
+  symex_heuristic_transformationt default_transformation;
+
+  /// The max complexity rating that a branch can be before it's abandoned. The
+  /// internal representation for computing complexity, has no relation to the
+  /// argument passed in via options.
+  std::size_t max_complexity = 0;
+
+  /// The amount of branches that can fail within the scope of a loops execution
+  /// before the entire loop is abandoned.
+  std::size_t max_loops_complexity = 0;
+
+  /// Checks whether the current loop execution stack has violated
+  /// max_loops_complexity.
+  bool are_loop_children_too_complicated(call_stackt &current_call_stack);
+
+  /// Checks whether we're in a loop that is currently considered blacklisted,
+  /// and shouldn't be executed.
+  static bool
+  in_blacklisted_loop(call_stackt &current_call_stack, const irep_idt &loop_id);
+
+  /// Returns inner-most currently active loop. This is frame-agnostic, so if
+  /// we're in a loop further up the stack that will still be returned as
+  /// the 'active' one.
+  static optionalt<loop_complexity_infot>
+  get_current_active_loop(call_stackt &current_call_stack);
```

Comment:
can it really mutate the currect_call_stack? :duck:

Reply:
Yeah, it grabs a pointer and mutates it later.